ime/app/src/main/java/com/anysoftkeyboard/ime/AnySoftKeyboardSuggestions.java

```diff
@@ -1236,10 +1236,34 @@ public void pickSuggestionManually(
                     suggestion,
                     suggestion /*user physically picked a word from the suggestions strip. this is not a fix*/);
 
-            // Follow it with a space
+            // Follow it with a space if there is not already one or if it is not a punctuation mark
+            // that goes attached to the word being manually picked
             if (withAutoSpaceEnabled && (index == 0 || !typedWord.isAtTagsSearchState())) {
-                sendKeyChar((char) KeyCodes.SPACE);
-                setSpaceTimeStamp(true);
+                boolean isNextCharSpaceOrPunctuation = false;
+                if (ic != null) {
+                    String strNextChar = ic.getTextAfterCursor(1, 0).toString();
```

Comment:
calling `getTextAfterCursor` is a very expensive call (it needs to do inter-process-communication between AnySoftKeyboard and the hosting app, and wait for both to have an idle main-thread). I suggest finding another way to get the last typed character - maybe store the `primaryCode` in a member field?

Reply:
actually, I'm not sure that would help, since you want to fix the case where the cursor was moved, too. Maybe, in that case (when `restart suggestions` is enabled) we should call `getTextAfterCursor` after the restart was done.